### DISCUSSION TITLE
Distribution Package with two Import Packages: Install only one Import Package

### DISCUSSION THREAD

**Discussion by buhtz** [pdm-project/pdm#1599] [General]
This question was also [asked on `setuptools` dicussion forum](https://github.com/pypa/setuptools/discussions/3749).
I know nothing then `setuptools` and now try to look around for alternatives that maybe able to solve my "problem".

Your package looks very [powerfull and flexible](https://pdm.fming.dev/latest/pyproject/build/). Maybe the use of [environment variables](https://pdm.fming.dev/latest/pyproject/pep621/#environment-variables) or other features can can help me with my problem?

I have a _Distribution Package_ named [`bit_demo`](https://codeberg.org/buhtz/bit_demo) containing the two _Import Packages_ `bitcli` and `bitgui`.

```
bit_demo
├── pyproject.toml
├── src
│   ├── bitcli
│   │   ├── __init__.py
│   │   └── __main__.py
│   └── bitgui
│       ├── __init__.py
│       └── __main__.py
└── tests
    ├── integration
    └── unit
```

When I install it via `python3 -m pip install bit_demo` I do have the two import packages in my system.

Would it be possible to install `bit_demo` but only with `bitcli`?
Can I install a _distribution package_ excluding some of its _import packages_?

For example

    python3 -m pip install bit_demo.bitcli

    python3 -m pip install bit_demo  --exclude bit_demo.bitgui

### Background of my question
I think about the distro maintainers (e.g. Debian, Arch) who need to create a deb file of my upstream repo.
There  need to be two packages in the distro. When users only want to use the CLI interface without GUI (e.g. on a headless server) they would do `apt install bitcli`. But when they need it with GUI they would do `apt install bitgui` (and `bitcli` comes in as dependency).

Technically `bitcli` can live alone. But `bitgui` is useless alone because it use a lot of `bitcli` code. Because of that I don't think it would be a good idea to create two separate git repositories. I would prefer to keep them together as one repo/distribution package.

**Comment by pawamoy** [comment]
There's no opt-out behavior baked in Python packaging. But there is opt-in:

```
pip install bit-demo[gui]
```

For this, you need to declare the `gui` extra for your `bit-demo` package. Search for "extra" in PDM's docs :)

Another solution would be to distribute your two packages as namespace packages. This would allow to:

```
pip install bit-demo.bitcli
pip install bit-demo.bitgui
```

...and maybe have a meta distribution that depends on both so you can

```
pip install bit-demo
```

**Reply by buhtz** [reply]
Never heard of _namespace packages_ before. Thanks for that hint.
But how many repos and pyproject.toml files do I need to realize something like `pip install bit-demo.bitcli` and `pip install bit-demo.bitgui`? Can I have all this in one repo only? btw: I (and my distro maintainers) don't install from PyPi but from to local cloned git repo.

**Reply by pawamoy** [reply]
Yeah that's probably the hard part :thinking:
You'd need two different `pyproject.toml` files I guess. Or two different folders. So you'd essentially either have a monorepo, or two different repositories.

About namespace packages, see the docs: https://packaging.python.org/en/latest/guides/packaging-namespace-packages/ (particularly "native namespace packages").

Not sure what commands you and your distro maintainers run once inside the local git clone. But there's probably a way, even with a monorepo.

**Comment by frostming** [comment]
> Your package looks very [powerfull and flexible](https://pdm.fming.dev/latest/pyproject/build/).

True, but not beyond the packaging standards 🤣
